extmod/modopenamp.c

```diff
@@ -0,0 +1,394 @@
+#define VRING_RX_ADDR           (METAL_SHM_ADDR)
+#define VRING_TX_ADDR           (METAL_SHM_ADDR + 0x1000)
+#define VRING_BUFF_ADDR         (METAL_SHM_ADDR + 0x2000)
+#define VRING_BUFF_SIZE         (METAL_SHM_SIZE - 0x2000)
+
+static const char openamp_trace_buf[128];
+#define MICROPY_PY_OPENAMP_TRACE_BUF       ((uint32_t)openamp_trace_buf)
+#define MICROPY_PY_OPENAMP_TRACE_BUF_LEN   sizeof(MICROPY_PY_OPENAMP_TRACE_BUF)
+
+#endif // MICROPY_PY_OPENAMP_RSC_TABLE_ENABLE
+
+#define debug_printf(...)       // mp_printf(&mp_plat_print, __VA_ARGS__)
+
+#if MICROPY_PY_OPENAMP_RPROC
+extern mp_obj_type_t rproc_type;
+#endif
+
+static struct metal_device shm_device = {
+    .name = METAL_SHM_NAME,
+    // The number of IO regions is fixed and must match the number and
+    // layout of the remote processor's IO regions. The first region is
+    // used for the vring shared memory, and the second one is used for
+    // the shared resource table.
+    .num_regions = METAL_MAX_DEVICE_REGIONS,
+    .regions = { { 0 } },
+    .node = { NULL },
+    .irq_num = 0,
+    .irq_info = NULL
+};
+static metal_phys_addr_t shm_physmap[] = { 0 };
+
+// ###################### Virtio device class ######################
+typedef struct _virtio_dev_obj_t {
+    mp_obj_base_t base;
+    struct rpmsg_virtio_device rvdev;
+    struct rpmsg_virtio_shm_pool shm_pool;
+    mp_obj_t ns_callback;
+} virtio_dev_obj_t;
+
+STATIC mp_obj_t virtio_dev_deinit(mp_obj_t self_in) {
+    virtio_dev_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    rpmsg_deinit_vdev(&self->rvdev);
+    metal_finish();
+    MP_STATE_PORT(virtio_device) = NULL;
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(virtio_dev_deinit_obj, virtio_dev_deinit);
+
+STATIC const mp_rom_map_elem_t virtio_dev_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_virtio_dev) },
+    { MP_ROM_QSTR(MP_QSTR___del__), MP_ROM_PTR(&virtio_dev_deinit_obj) },
+};
+STATIC MP_DEFINE_CONST_DICT(virtio_dev_locals_dict, virtio_dev_locals_dict_table);
+
+MP_DEFINE_CONST_OBJ_TYPE(
+    virtio_dev_type,
+    MP_QSTR_virtio_dev,
+    MP_TYPE_FLAG_NONE,
+    locals_dict, &virtio_dev_locals_dict
+    );
+
+// ###################### RPMsg Endpoint class ######################
+typedef struct _rpmsg_obj_t {
+    mp_obj_base_t base;
+    mp_obj_t name;
+    mp_obj_t callback;
+    struct rpmsg_endpoint ep;
+} rpmsg_obj_t;
+
+const mp_obj_type_t rpmsg_type;
+
+static int rpmsg_recv_callback(struct rpmsg_endpoint *ept, void *data, size_t len, uint32_t src, void *priv) {
+    debug_printf("rpmsg_recv_callback() message received src_addr: %lu msg len: %d\n", src, len);
+    rpmsg_obj_t *self = metal_container_of(ept, rpmsg_obj_t, ep);
+    if (self->callback != mp_const_none) {
+        mp_call_function_2(self->callback, mp_obj_new_int(src), mp_obj_new_bytearray_by_ref(len, data));
+    }
+    return 0;
+}
+
+STATIC mp_obj_t _rpmsg_send(uint n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
```

Comment:
Why does this need the leading underscore? I'm guessing it's conflicting with a symbol in OpenAMP. Technically leading underscore names are reserved, so we avoid using them.

Perhaps we should prefix all the symbols in this file with `modopenamp_` so that everything is consistent.